### PR TITLE
MRG: Better mesh options

### PR DESCRIPTION
**Pull request by larsoner**
Allows doing a vertex value->colormap mapping with meshes.

### CODE REVIEW

File: vispy/visuals/mesh.py [hunk 1]
```diff
@@ -28,12 +29,10 @@
 varying vec4 v_light_color;
 varying vec4 v_base_color;
 
-
 void main() {
-
     v_ambientk = $ambientk;
     v_light_color = $light_color;
-    v_base_color = $base_color;
+    v_base_color = $color_transform($base_color);
```
Comment (Eric89GXL) [inline vispy/visuals/mesh.py]: so here before we used to just use the variable `$base_color`, but I needed to add a color transform on top of that to use the colormap mapping

File: vispy/visuals/mesh.py [hunk 2]
```diff
@@ -136,6 +127,21 @@
...
+
+def _build_color_transform(data, cmap, clim=(0., 1.)):
+    if data.ndim == 2 and data.shape[1] == 1:
+        fun = Function(_clim)
+        fun['cmin'] = clim[0]
+        fun['cmax'] = clim[1]
+        fun = FunctionChain(None, [fun, Function(cmap.glsl_map)])
+    else:
+        fun = Function(_null_color_transform)
+    return fun
```
Comment (Eric89GXL) [inline vispy/visuals/mesh.py]: These are the bits of Python code necessary to build the `$color_transform` variable above. If it's not a colormap, the `_null_*` is used, which is just a pass-through. If it is a colormap, then we chain the colormap's `glsl_map` (which is GLSL code) after applying some simlpe GLSL code to set upper and lower limits to the data (`clim`).

File: vispy/visuals/mesh.py [hunk 3]
```diff
@@ -357,16 +424,77 @@ def _update_data(self):
...
+        ----------
+        direction : ndarray, shape (3,)
+            The light direction.
+        """
+        direction = np.array(direction, float).ravel()
+        if direction.size != 3 or not np.isfinite(direction).all():
+            raise ValueError('Invalid direction %s' % direction)
+        self._light_dir = tuple(direction)
+        self.mesh_data_changed()
+
```
Comment (Eric89GXL) [inline vispy/visuals/mesh.py]: All of these other changes are merely to expose some parameters that were formerly hard-coded.

File: vispy/visuals/mesh.py [hunk 4]
```diff
@@ -321,13 +386,17 @@ def _update_data(self):
...
+                colors = colors.astype(np.float32)
             elif md.has_face_color():
-                self._colors.set_data(md.get_face_colors(indexed='faces'),
-                                      convert=True)
+                colors = md.get_face_colors(indexed='faces')
+                colors = colors.astype(np.float32)
+            elif md.has_vertex_value():
+                colors = md.get_vertex_values(indexed='faces')
+                colors = colors.ravel()[:, np.newaxis]
+                colors = colors.astype(np.float32)
```
Comment (Eric89GXL) [inline vispy/visuals/mesh.py]: oh, and of course we need to add a conditional that causes it to actually set `colors` when the vertex-value mode is available

File: vispy/visuals/mesh.py [hunk 5]
```diff
@@ -357,16 +424,77 @@ def _update_data(self):
                 normals = (1., 0., 0.)
 
             self.shared_program.vert['normal'] = normals
-            self.shared_program.vert['base_color'] = colors
+            self.shared_program.vert['color_transform'] = \
+                _build_color_transform(colors, self._cmap, self._clim_values)
+            if colors.ndim == 1:
+                self.shared_program.vert['base_color'] = colors
+            else:
+                self.shared_program.vert['base_color'] = VertexBuffer(colors)
```
Comment (Eric89GXL) [inline vispy/visuals/mesh.py]: and a conditional based on whether we're doing the value->color conversion with colormap, or the old/standard way (per-vertex colors actually passed in)

**Comment by larsoner** [comment]
This is ready for review/merge from my end

File: vispy/gloo/buffer.py [hunk 6]
```diff
@@ -191,7 +191,13 @@ def set_data(self, data, copy=False, **kwargs):
...
         data = self._prepare_data(data, **kwargs)
         self._dtype = data.dtype
-        self._stride = data.strides[-1]
+        # This works around some strange NumPy bug where a float32 array
+        # of shape (155407, 1) was said to have strides
+        # (4, 9223372036854775807), which is crazy
+        if data.ndim == 2 and data.shape[1] == 1:
+            self._stride = data.strides[0]
+        else:
+            self._stride = data.strides[-1]
```
Comment (davidh-ssec) [inline vispy/gloo/buffer.py]: Any idea if this bug still exists?

Comment (larsoner) [reply vispy/gloo/buffer.py]: Perhaps it was a bleeding-edge NumPy bug. I don't see it anymore so I've removed it

File: vispy/visuals/mesh.py [hunk 7]
```diff
@@ -136,6 +131,21 @@
 """)
 
 
+_null_color_transform = 'vec4 pass(vec4 color) { return color; }'
+_clim = 'float cmap(float val) { return (val - $cmin) / ($cmax - $cmin); }'
+
+
+def _build_color_transform(data, cmap, clim=(0., 1.)):
```
Comment (davidh-ssec) [inline vispy/visuals/mesh.py]: Is it worth it to try to combine this `_build_color_transform` with the one in `image.py`? I know in one of my own projects that uses vispy I end up using clims with an ImageVisual subclass.

Comment (larsoner) [reply vispy/visuals/mesh.py]: It looks like `visuals/image.py` (line ~92) uses a transformation of `(color.r + color.g + color.b / 3.)` whereas this just takes and converts a `float`. So it's not quite so straightforward unfortunately. I'll add a comment that eventually they could be unified, though.

**Pull request merged**